### PR TITLE
New Morphology methods and improvements

### PR DESCRIPTION
**Pull request by Idclip**
Aims of this PR:
 - Deprecate `dilateVoxels` and `erodeVoxels` in favour of `dilateActiveLeafValues` and `erodeActiveLeafValues`
 - Introduce `erodeActiveValues` to match `dialteActiveValues` to handle tile policies for erosion
 - Introduce multithreaded erosion of edge and vertex neighbours
 - Significantly improve the performance of large dilations
 - Remove unused code and significantly clean-up Morphology.h such that namespaced classes can be used by clients for more granular control

NOTE:
 - Incorporates PR #751
 - `erodeVoxels` calls `pruneLevelSet`, the new methods do **not** and leave such decisions either to the tile policy or client code.

As of this PR `dilateVoxels` simply called `dilateActiveLeafValues` (unthreaded) and `erodeVoxels` calls `erodeActiveLeafValues`, but follows with a call to `pruneLevelSet`. The behaviour of these old functions thus remains exactly the same.

Results for discussion:
https://docs.google.com/spreadsheets/d/13-6yTmmPdj_gLLDj-3hgaCUaDy_-hgs-aY1BoMQr6PQ/edit?usp=sharing

Signed-off-by: Nick Avramoussis <4256455+Idclip@users.noreply.github.com>

### CODE REVIEW

File: openvdb/tools/Activation.h
Comment (Idclip) [inline openvdb/tools/Activation.h:9]: This file is a copy/paste of the activation/deactivation methods out of Morphology.h and can be mostly be ignored

File: openvdb_houdini/SOP_OpenVDB_Activate.cc
Comment (Idclip) [inline openvdb_houdini/SOP_OpenVDB_Activate.cc:509]: @jmlait wasn't too sure what to do here. The VDB Activate SOP used to call `erodeVoxels` which always calls `pruneLevelSet`. If the grid is not marked as a level set (but really, it's if the grid has a negative background value), the VDB Activate SOP will fail with an error if you try to expand with negative values (i.e. erode). To solve this and maintain compatible behaviour, I've just branched on the grid type. However...this SOP also has a "Prune Tolerance" parameter which also calls "prune". I Could remove this branching in favour of that parameter, but it will change behaviour for anyone who is eroding and has the "Prune Tolerance" disabled.

Another question, expansion currently only expands leaf nodes (due to the old impl of dilateVoxels and erodeVoxels). I could update this to handle tiles with the new methods now? Again, it would change behaviour.

File: openvdb/openvdb/tools/Morphology.h
Comment (jmlait) [inline openvdb/openvdb/tools/Morphology.h:335]: Why not assert(false) and do nothing in this case?

**Review by jmlait — APPROVED** [review]

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:335]: Yeah valid point, will change

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Activate.cc
Comment (Idclip) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Activate.cc:509]: @jmlait just an FYI, I have not changed the behaviour of the activate SOP in this PR; it still only works on voxels

**Comment by Idclip** [comment]
I have a single commit way back when that wasn't signed off which is extremely annoying. As we can't use the DCO signoff button, I'll have to remake this PR, but I will do that once it's been approve by 1 more person

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [inline openvdb/openvdb/tools/Morphology.h:158]: I remember talking about this but I can't remember the reason. I still find it rather odd that we have `dilateActiveValues(..., IGNORE_TILES)` and `dilateActiveLeafValues(...)` which are the same thing?

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [inline openvdb/openvdb/tools/Morphology.h:245]: Does this `preserveMaskLeafNodes` value actually get used anywhere? I can't seem to find any references to it being true.

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [inline openvdb/openvdb/tools/Morphology.h:307]: It confused me this was called `cache` not `mask`.

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [inline openvdb/openvdb/tools/Morphology.h:638]: Did you try this without using a LeafManager - ie just a linear list of nodes with a on/off toggle for each one? Feels like you might be spending some time extracting and re-inserting leaf nodes and rebuilding the leaf manager with this technique?

**Review by danrbailey — COMMENTED** [review]
In general looks good. A few additional questions:

* Should there also be a unit test for the new activation header?
* You presented some nice benchmark graphs of the results of this PR - are these included in the unit tests now?
* Many of the parallel_for methods you're using do very simple operations that I imagine is only a handful of instructions, did you consider using a larger grain size than 1?
* I would expect the non-threaded mode to call a lambda directly rather than to call into tbb::parallel_for() with a very large grain size. I think we're fairly consistent about how we do that, would be nice to avoid the tbb initialization if it's not strictly necessary. I also find it tends to make it easier to track down an issue because the stracktraces from tbb can be a bit verbose.

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:256]: >  Have you got data to back this up? 

Yes of course... I did anyway. I encourage you to run your own tests as it's only noticeable with smaller operations, most likely due to the very minor overhead of extra data on the internal leaf iterator sub class.

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:158]: I think this was to provide an easier 1:1 mapping from the old functions. I guess I could remove them?

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:245]: No it doesn't. This was introduced due a problem with the underlying implementation. This algorithm uses a mask topology  to perform the dilation. If the input is already a mask topology it just uses it directly. When dilating the mask topology, it steals the nodes and splits them across active threads. When it comes to re-combine the nodes, it's possible it'll favour newly allocated data over the input nodes, resulting in any held leaf pointers being invalidated. Users can use this internal option to retain the state for input mask dilation. It's impossible to know if this is required or not as the alternative is to deep copy the mask tree, a significant overhead. It's good to have this as an option (I think a problem with various bits of the VDB API is that it's sometimes clear what guarantees are provided from various functions w.r.t the state of the tree).

**Comment by Idclip** [comment]
@danrbailey 

> * Should there also be a unit test for the new activation header?

Yes? But I'm not going to add it in this PR. I didn't touch this code, it's simply been moved out of Morphology.h.

> * You presented some nice benchmark graphs of the results of this PR - are these included in the unit tests now?

No, you can see on the spreadsheets that the VDBs in use were Houdini test assets.

> * Many of the parallel_for methods you're using do very simple operations that I imagine is only a handful of instructions, did you consider using a larger grain size than 1?

No, I can do another pass.

> * I would expect the non-threaded mode to call a lambda directly rather than to call into tbb::parallel_for() with a very large grain size.

As would I, I will change these.

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:638]: For iter>1 I always need to call `getNodes`. So unless `initLeafArray` is slower than `getNodes`, the leaf manager way is better.

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: renamed

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: done

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: done

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: changed

File: CHANGES
Comment (Idclip) [reply CHANGES]: renamed

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: fixed

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [reply openvdb/openvdb/tools/Morphology.h:158]: I'm be in favor of removing `dilateActiveLeafValues`. Perhaps a little more work to migrate, but less API to support is good?

File: openvdb/openvdb/tools/Morphology.h
Comment (danrbailey) [reply openvdb/openvdb/tools/Morphology.h:245]: I don't mind leaving this flag in, but I think if we include it without ever actually using it ourselves, it should at least be evaluated in the unit tests, which I don't think it is currently?

**Comment by danrbailey** [comment]
> @danrbailey
> 
> > * Should there also be a unit test for the new activation header?
> 
> Yes? But I'm not going to add it in this PR. I didn't touch this code, it's simply been moved out of Morphology.h.

Ok, that's fine. Perhaps I can add this when I look at switching out the implementation of the VDB Activate SOP.

> > * You presented some nice benchmark graphs of the results of this PR - are these included in the unit tests now?
> 
> No, you can see on the spreadsheets that the VDBs in use were Houdini test assets.

If you happen to have it still, would be nice to email around the hip file with these tests in it. If not, no worries.

> > * Many of the parallel_for methods you're using do very simple operations that I imagine is only a handful of instructions, did you consider using a larger grain size than 1?
> 
> No, I can do another pass.
> 
> > * I would expect the non-threaded mode to call a lambda directly rather than to call into tbb::parallel_for() with a very large grain size.
> 
> As would I, I will change these.

Great, thanks for the other changes.

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h]: Removed both `dilateActiveLeafValues` and `erodeActiveLeafValues`

File: openvdb/openvdb/tools/Morphology.h
Comment (Idclip) [reply openvdb/openvdb/tools/Morphology.h:205]: Added a test

**Comment by Idclip** [comment]
> If you happen to have it still, would be nice to email around the hip file with these tests in it. If not, no worries.

I have the scene and will e-mail it around soon.

I think I've addressed all feedback. A reminder that once this is approved I'll have to close and re-create these changes on a new PR because of a single commit missing dco signoff and being told not to use the nice dco button :|

**Review by danrbailey — APPROVED** [review]
This is great, thanks for making those changes and emailing around the scene, Nick. I'm happy to approve the new PR too.

**Comment by danrbailey** [comment]
I think we should bump the next release to 8.1 once this PR gets merged as well as there are some API changes included in this PR.

**Comment by Idclip** [comment]
Awesome. thanks all. I've remade the PR here: https://github.com/AcademySoftwareFoundation/openvdb/pull/973

Closing this due to DCO issues. PR https://github.com/AcademySoftwareFoundation/openvdb/pull/973 is completely identical but without the history.

**Pull request closed without merging**
